## Notebook match and combine dataframes with the enrollment data

### going to use fuzzy matching to acomplish the task

In [162]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### File Locations
# File location and type

# 2022 Enrollment Data
enroll_df = pd.read_excel('data\collected.xlsx', sheet_name='2022_enrollment')

# Sheet with manually collected data
manual_df = pd.read_excel('data\collected.xlsx', sheet_name='manual')

# Sheet with calculated data
area_sqft_df = pd.read_excel('data\collected.xlsx', sheet_name='area_sqft')

high_school_df = manual_df[manual_df['Level'] == 'high school']
# high_school_df.head()
# enroll_df.head()
# area_sqft_df.head()

##### Working to Load all the data



In [163]:
### Lookup enrollment numbers and school id from the enrollment sheet and add them to the high school sheet

import difflib

high_school_df['match'] = high_school_df['School'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_name'].tolist(), n=1, cutoff=0.8))
high_school_df['match'] = high_school_df['match'].apply(lambda x: x[0] if len(x) > 0 else np.nan)
high_school_df['enrollment'] = high_school_df['match'].apply(lambda x: enroll_df[enroll_df['school_name'] == x]['enrollment_classification'].values[0] if type(x) == str else np.nan)
high_school_df['school_id'] = high_school_df['match'].apply(lambda x: enroll_df[enroll_df['school_name'] == x]['school_id'].values[0] if type(x) == str else np.nan)
high_school_df['Div'] = high_school_df['match'].apply(lambda x: enroll_df[enroll_df['school_name'] == x]['division'].values[0] if type(x) == str else np.nan)
high_school_df['enrollment'] = high_school_df['enrollment'].fillna(0)
high_school_df['school_id'] = high_school_df['school_id'].fillna(0)
high_school_df = high_school_df.drop(columns=['match'])
high_school_df.head()

high_school_df.describe()

## Export csv to check output

# high_school_df.to_csv('data\high_school_df_test.csv')



C:\Users\Justin\AppData\Local\Temp\ipykernel_904\253980905.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  high_school_df['match'] = high_school_df['School'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_name'].tolist(), n=1, cutoff=0.8))
C:\Users\Justin\AppData\Local\Temp\ipykernel_904\253980905.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  high_school_df['match'] = high_school_df['match'].apply(lambda x: x[0] if len(x) > 0 else np.nan)
C:\Users\Justin\AppData\Local\Temp\ipykerne

FOP_GE_area    FOP_perim  Foul_GE_area   Foul_perim  FOP_ratio  \
count     148.000000   147.000000    148.000000   148.000000  63.000000   
mean    90254.837838  1178.319728  27357.493243  1398.128378  76.227799   
std      8070.986379    55.351305   6528.582436    72.717072   4.195067   
min     70894.000000  1046.000000  13992.000000  1166.000000  62.560057   
25%     86184.250000  1150.500000  21988.500000  1355.750000  74.189485   
50%     90089.500000  1176.000000  27621.500000  1392.500000  76.318259   
75%     93751.750000  1200.500000  32033.250000  1431.500000  78.282507   
max    139904.000000  1540.000000  46286.000000  1750.000000  91.496469   

       Foul-ratio  Foul_pct_of_whole  YE_2022_rank  Region_div   enrollment  \
count   63.000000          63.000000     76.000000   60.000000   148.000000   
mean    19.887823           0.235314     10.328947    2.550000   337.209459   
std      4.503542           0.044588      5.881923    1.111191   558.265490   
min     10.629978           0.131112      1.000000    1.000000     0.000000   
25%     16.229598           0.207135      5.000000    2.000000     0.000000   
50%     20.007988           0.237437     10.000000    2.500000     0.000000   
75%     23.240048           0.267046     15.250000    4.000000   457.750000   
max     27.961942           0.313700     20.000000    4.000000  2812.000000   

         school_id  
count   148.000000  
mean   1991.844595  
std    2662.015997  
min       0.000000  
25%       0.000000  
50%       0.000000  
75%    3633.250000  
max    8192.000000

In [167]:
## Lookup enrollment numbers and school id from the enrollment sheet and add them to the area_sqft sheet

import difflib

area_sqft_df['match'] = area_sqft_df['field'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_name'].tolist(), n=1, cutoff=0.8))
area_sqft_df['match'] = area_sqft_df['match'].apply(lambda x: x[0] if len(x) > 0 else np.nan)
area_sqft_df['enrollment'] = area_sqft_df['match'].apply(lambda x: enroll_df[enroll_df['school_name'] == x]['enrollment_classification'].values[0] if type(x) == str else np.nan)
area_sqft_df['school_id'] = area_sqft_df['match'].apply(lambda x: enroll_df[enroll_df['school_name'] == x]['school_id'].values[0] if type(x) == str else np.nan)
area_sqft_df['Div'] = area_sqft_df['match'].apply(lambda x: enroll_df[enroll_df['school_name'] == x]['division'].values[0] if type(x) == str else np.nan)
area_sqft_df['enrollment'] = area_sqft_df['enrollment'].fillna(0)
area_sqft_df['school_id'] = area_sqft_df['school_id'].fillna(0)
area_sqft_df = area_sqft_df.drop(columns=['match'])
area_sqft_df.info()

## Export csv to check output
area_sqft_df.to_csv('area_sqft_df_test.csv')



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 222 entries, 0 to 221
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  222 non-null    int64  
 1   field       222 non-null    object 
 2   home_plate  222 non-null    object 
 3   foul_ft     222 non-null    float64
 4   fop_ft      222 non-null    float64
 5   foul_pct    222 non-null    float64
 6   enrollment  222 non-null    float64
 7   school_id   222 non-null    float64
 8   Div         64 non-null     object 
dtypes: float64(5), int64(1), object(3)
memory usage: 15.7+ KB


## Code above works

### the lookups need to be a lot better. only matched 64 records out of 222


#### the problem might just be how messy the field names are in the area_sqft_df
#### might need to clean them up at the google maps level

In [151]:
##### THIS BLOCK IS KIND OF A MESS AND PROBABLY ISN'T GOING TO BE USED

### merge the dataframes
import difflib
# from fuzzywuzzy import fuzz
# from fuzzywuzzy import process


high_school_df['match'] = high_school_df['School'].str.lower()
enroll_df['match'] = enroll_df['school_name'].str.lower()
area_sqft_df['match'] = area_sqft_df['field'].str.lower()

## Drop the columns that are not needed
high_school_df = high_school_df.drop(columns=['FOP_ratio','Foul-ratio','Foul_pct_of_whole'])


## Match the school names
high_school_df['mat2'] = high_school_df['match'].apply(lambda x: difflib.get_close_matches(x, enroll_df['match'], n=1, cutoff=0.8))
area_sqft_df['mat2'] = area_sqft_df['match'].apply(lambda x: difflib.get_close_matches(x, enroll_df['match'], n=1, cutoff=0.8))
enroll_df['mat2'] = high_school_df['match'].apply(lambda x: difflib.get_close_matches(x, enroll_df['match'], n=1, cutoff=0.8))

enroll_df['mat2'] = enroll_df['mat2'].to_string()
high_school_df['mat2'] = high_school_df['mat2'].to_string()
area_sqft_df['mat2'] = area_sqft_df['mat2'].to_string()
temp_df = high_school_df.merge(area_sqft_df, on='mat2', how = 'outer')
# new_df = temp_df.merge(enroll_df, on='mat2',how='right')

temp_df.head()

#### Merge seems to work
## Output a csv to see what is going on
# temp_df.to_csv('new_df.csv')


## cAN'T make the merge work yet
# outputing the dataframes to csv to see what is going on
# high_school_df.to_csv('high_school_df.csv')
# enroll_df.to_csv('enroll_df.csv')
area_sqft_df.to_csv('area_sqft_df.csv')

# ## Merge the dataframes
# new_df = high_school_df.merge(area_sqft_df, on='mat2',how='right')

# new_df.head()
# high_school_df.head()
# enroll_df.head()
# area_sqft_df.head()

C:\Users\Justin\AppData\Local\Temp\ipykernel_904\1073604786.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  high_school_df['match'] = high_school_df['School'].str.lower()


In [85]:
import difflib 

# high_school_df['school_match'] = high_school_df['School'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_name']))
# # high_school_df['school_match'] = high_school_df['school_match']
# high_school_df.dtypes

# new_df = high_school_df.join(enroll_df, on='school_name', how='left', lsuffix='_left', rsuffix='_right')
# new_df
enroll_df['school_match'] = enroll_df['school_name']

area_sqft_df['school_match'] = area_sqft_df['field'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_match']))
new_df = area_sqft_df.join(enroll_df, on='school_match', how='left', lsuffix='_left', rsuffix='_right')

# area_sqft_df['school_match'] = area_sqft_df['school_match']
# area_sqft_df.dtypes

high_school_df.dtypes
# area_sqft_df.dtypes


# new_df = area_sqft_df.join(high_school_df, on='school_match', how='left', lsuffix='_left', rsuffix='_right')
# new_df


# # Merge the dataframes
# out_df = pd.merge(high_school_df, area_sqft_df, on='school_match', how='left')
# out_df



# area_sqft_df['school_match'] = area_sqft_df['field'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_match']))
# area_sqft_df
# high_school_df.to_csv('data\high_school_df.csv')

# enroll_df['school_match'] = enroll_df['school_name'].apply(lambda x: difflib.get_close_matches(x, high_school_df['School']))
# enroll_df
# #create duplicate column to retain team name from df2
# high_school_df['team'] = high_school_df['School']

# #convert team name in df2 to team name it most closely matches in df1
# high_school_df['school_name'] = high_school_df['team'].apply(lambda x: difflib.get_close_matches(x, enroll_df['school_name']))

# #merge the DataFrames into one
# out_df = pd.merge(high_school_df, enroll_df, on='school_name', how='left')

# #view final DataFrame
# print(out_df)

ValueError: You are trying to merge on object and int64 columns. If you wish to proceed you should use pd.concat